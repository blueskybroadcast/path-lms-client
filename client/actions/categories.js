import normalize from 'jsonapi-normalizer';

import { FETCH_CATEGORIES, DELETE_CATEGORY } from './types';

export const fetchCategories = () => async (dispatch, getState, api) => {
  const res = await api.get('/categories');

  dispatch({
    type: FETCH_CATEGORIES,
    payload: normalize(res.data)
  });
};

export const addCategory = category => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.post('/categories', {
    name,
    parent_id: parentId,
    hidden
  }).then(() => dispatch(fetchCategories()));
};

export const deleteCategory = id => async (dispatch, getState, api) => {
  await api.delete(`/categories/${id}`).then(() => {
    dispatch({
      type: DELETE_CATEGORY,
      payload: id
    });
  });
};

export const editCategory = (id, category) => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.put(`/categories/${id}`, {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch(fetchCategories());
  });
};
