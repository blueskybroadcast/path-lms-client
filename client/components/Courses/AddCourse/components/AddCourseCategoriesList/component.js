import React from 'react';
import PropTypes from 'prop-types';

const AddCourseCategoriesList = ({ categories, categoryIds, handleCategoryToggle }) => (
  <ul>
    {categories.map((category) => {
      const catId = category.id;
      const catIndex = categoryIds.indexOf(catId);
      const isChecked = catIndex !== -1;
      return (
        <li
          className={`level-${category.level}`}
          key={catId}
        >
          <label>
            <input
              checked={isChecked}
              onChange={() => handleCategoryToggle(catId, catIndex, isChecked)}
              type="checkbox"
            />
            {category.name}
          </label>
        </li>
      );
    })}
  </ul>
);

AddCourseCategoriesList.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.object).isRequired,
  categoryIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleCategoryToggle: PropTypes.func.isRequired
};

export default AddCourseCategoriesList;
