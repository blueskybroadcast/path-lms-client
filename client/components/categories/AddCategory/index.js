import AddCategory from './container';

export default AddCategory;
