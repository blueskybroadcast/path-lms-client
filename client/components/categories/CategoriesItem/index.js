import CategoriesItem from './container';

export default CategoriesItem;
