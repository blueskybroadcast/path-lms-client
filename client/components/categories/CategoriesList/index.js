import CategoriesList from './container';

export default CategoriesList;
