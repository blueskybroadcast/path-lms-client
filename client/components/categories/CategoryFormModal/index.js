import CategoryFormModal from './container';

export default CategoryFormModal;
