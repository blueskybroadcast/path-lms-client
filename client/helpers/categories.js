export const getParentListLevelOne = (categories) => {
  const parentsLevelOne = categories.filter(cat => cat.level === 0 && cat.childCount > 0);
  return parentsLevelOne.map(({ name, id }) => ({
    name,
    list: categories.filter(cat => cat.parentId === id)
  }));
};

export const getSortedCategories = (categoriesIds, categoriesData) => {
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        level: 0,
        name: data.name,
        childCount: 0,
        hidden: data.hidden
      });
    }
  });

  const nextCategories = [...categories];
  let nextAddedCategoriesCount = 0;

  categories.forEach((cat, index) => {
    categoriesIds.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        nextCategories.splice(
          index + 1 + nextAddedCategoriesCount,
          0,
          {
            id,
            level: 1,
            name: data.name,
            childCount: 0,
            parentId,
            hidden: data.hidden
          }
        );
        categories[index].childCount += 1;
        nextAddedCategoriesCount += 1;
      }
    });
  });

  const finalCategories = [...nextCategories];
  let finalAddedCategoriesCount = 0;

  nextCategories.forEach((cat, index) => {
    if (cat.level === 1) {
      categoriesIds.forEach((id) => {
        const data = categoriesData[id];
        if (cat.id === data.parent.id) {
          finalCategories.splice(
            index + 1 + finalAddedCategoriesCount,
            0,
            {
              id,
              level: 2,
              name: data.name,
              hidden: data.hidden
            }
          );
          nextCategories[index].childCount += 1;
          finalAddedCategoriesCount += 1;
        }
      });
    }
  });

  return finalCategories;
};

export const getCategoriesForSelect = (categoriesIds, categoriesData) => {
  if (!categoriesIds.length) return [];
  const remainingCategories = [...categoriesIds];
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        name: data.name,
        hidden: data.hidden,
        children: []
      });
      const itemToRemoveIndex = remainingCategories.indexOf(id);
      remainingCategories.splice(itemToRemoveIndex, 1);
    }
  });

  const finalCategories = [...categories];

  categories.forEach((cat, index) => {
    remainingCategories.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        const levelOneCat = {
          id,
          name: data.name,
          hidden: data.hidden,
          children: []
        };

        remainingCategories.forEach((childId) => {
          const childData = categoriesData[childId];
          const childParentId = childData.parent.id;

          if (id === childParentId) {
            const levelTwoCat = {
              id: childId,
              name: childData.name,
              hidden: childData.hidden,
              children: []
            };

            levelOneCat.children.push(levelTwoCat);
          }
        });

        finalCategories[index].children.push(levelOneCat);
      }
    });
  });

  return finalCategories;
};
