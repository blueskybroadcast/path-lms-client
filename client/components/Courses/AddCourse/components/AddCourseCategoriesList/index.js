import AddCourseCategoriesList from './component';

export default AddCourseCategoriesList;
