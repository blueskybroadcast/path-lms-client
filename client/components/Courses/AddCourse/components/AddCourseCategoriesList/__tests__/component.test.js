import React from 'react';
import { shallow } from 'enzyme';
import AddCourseCategoriesList from '../component';

import { visibleCategories } from '../../../../../../fixtures/categories';

describe('AddCourseCategoriesList', () => {
  test('should render AddCourseCategoriesList with data correctly', () => {
    const wrapper = shallow(
      <AddCourseCategoriesList
        categories={visibleCategories}
        categoryIds={['32', '33']}
        handleCategoryToggle={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render AddCourseCategoriesList without data correctly', () => {
    const wrapper = shallow(
      <AddCourseCategoriesList
        categories={[]}
        categoryIds={[]}
        handleCategoryToggle={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
