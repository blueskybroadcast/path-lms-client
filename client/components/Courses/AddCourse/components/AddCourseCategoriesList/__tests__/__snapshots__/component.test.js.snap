// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCourseCategoriesList should render AddCourseCategoriesList with data correctly 1`] = `
<ul>
  <li
    className="level-0"
    key="36"
  >
    <label>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      sda
    </label>
  </li>
  <li
    className="level-1"
    key="37"
  >
    <label>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      gffdsg
    </label>
  </li>
  <li
    className="level-0"
    key="32"
  >
    <label>
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      Test Test
    </label>
  </li>
  <li
    className="level-2"
    key="34"
  >
    <label>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      Tea sadasd s
    </label>
  </li>
</ul>
`;

exports[`AddCourseCategoriesList should render AddCourseCategoriesList without data correctly 1`] = `<ul />`;
